rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,375 @@
+    }
+
+  private def rehash(
+      path: KeyPath,
+      rest: Vector[Trie],
+      lastSeen: Trie,
+      acc: List[Trie]
+  ): F[List[Trie]] = {
+    type Params = (KeyPath, Vector[Trie], Trie, List[Trie])
+    def go(params: Params): F[Either[Params, List[Trie]]] = params match {
+      case (
+          currentPath: KeyPath,
+          currentRest: Vector[Trie],
+          nextLastSeen: Trie,
+          currentAcc: List[Trie]
+          ) =>
+        currentRest match {
+          case head :+ Node(ptr) =>
+            for {
+              pointerBlock <- pointerBlockStore
+                               .get(ptr)
+                               .map(_.get.updated((toInt(currentPath.last), nextLastSeen) :: Nil))
+              nodeToPB <- storePointerBlock(pointerBlock)
+            } yield (currentPath.init, head, nodeToPB, currentAcc :+ nextLastSeen).asLeft
+
+          case head :+ Skip(affix, _) =>
+            Applicative[F]
+              .pure(
+                (
+                  currentPath.dropRight(affix.size.toInt),
+                  head,
+                  Skip(affix, Trie.hash(nextLastSeen)),
+                  currentAcc :+ nextLastSeen
+                )
+              )
+              .map(_.asLeft)
+          case Vector() => Applicative[F].pure(currentAcc :+ nextLastSeen).map(_.asRight)
+        }
+    }
+    FlatMap[F].tailRecM((path, rest, lastSeen, acc))(go)
+  }
+
+  private def rebalanceDelete(p1: KeyPath, r: Vector[Trie], a: Option[Trie]): F[List[Trie]] = {
+    type Params = (KeyPath, Vector[Trie], Option[Trie])
+    def go(params: Params): F[Either[Params, List[Trie]]] =
+      params match {
+        case (currentKey, init :+ Skip(affix, _), None) => // remove empty skip
+          Applicative[F].pure((currentKey.dropRight(affix.size.toInt), init, None)).map(_.asLeft)
+
+        case (currentKey, init :+ Skip(affix, _), Some(Skip(eaffix, ep))) => // merge skips
+          rehash(currentKey.dropRight(affix.size.toInt), init, Skip(affix ++ eaffix, ep), Nil)
+            .map(_.asRight)
+
+        case (keyInit :+ keyLast, init :+ Node(p), None) => // handle pointer block
+          pointerBlockStore.get(p).flatMap {
+            case Some(pointerBlock) if pointerBlock.countNonEmpty == 2 =>
+              // pointerBlock contains 1 more element, find it, wrap in skip and re-balance
+              def getOther =
+                pointerBlock
+                  .updated((toInt(keyLast), EmptyTrie) :: Nil)
+                  .toVector
+                  .zipWithIndex
+                  .filter(v => v._1 != EmptyTrie)
+                  .head
+
+              val (other, idx) = getOther
+              Applicative[F].pure((keyInit, init, Some(pointTo(toByte(idx), other)))).map(_.asLeft)
+            case Some(pb) => // pb contains 3+ elements, drop one, rehash
+              val updated = pb.updated((toInt(keyLast), EmptyTrie) :: Nil)
+              storePointerBlock(updated)
```

Comment:
why do we `storePointerBlock` here, but not in the previous case? Both cases update the pointerBlock in question

Reply:
also, why doesn't the previous case need a rehash?

Reply:
because it's not done. the returned Left will continue the loop